zproject/backends.py

```diff
@@ -868,6 +883,61 @@ def sync_custom_profile_fields_from_ldap(
         except SyncUserError as e:
             raise ZulipLDAPError(str(e)) from e
 
+    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
+        if user_profile.realm.id not in settings.LDAP_GROUP_SYNC_LIST:
+            return  # no groups synced for this realm
+
+        ldap_group_sync_list = set(settings.LDAP_GROUP_SYNC_LIST[user_profile.realm.id])
+
+        try:
+            ldap_logger.debug("found groups for user: %s", str(user_profile))
+            ldap_group_name_set = set(ldap_user.group_names).intersection(ldap_group_sync_list)
+
+            member_group_name_set = {
+                membership.user_group.name
+                for membership in UserGroupMembership.objects.filter(
+                    user_group__realm=user_profile.realm,
+                    user_group__name__in=ldap_group_sync_list,
+                    user_profile=user_profile,
+                )
+            }
```

Comment:
This still does a lot of queries one-by-one in a loop (https://github.com/zulip/zulip/pull/23039#discussion_r1014911194) -  I think I didn't explain this well in that original comment. The idea is that first the `UserGroupMembership.objects.filter(...)` query is done, which is great, but then we loop over each `membership` and access `membership.user_group` - for each such access, Django has to go the database case the get the `UserGroup` from the database, which is very inefficient. The ugly query I've put in that comment was to grab the group names in a single query (since it's just the group names that we need in the end)